### PR TITLE
Adding measurement methods

### PR DESCRIPTION
**Pull request by hillmich**
The measurement methods are better suited to be in the DD package instead of the top level projects.

### CODE REVIEW

File: include/dd/Export.hpp [hunk 1]
```diff
@@ -654,6 +654,48 @@ namespace dd {
...
+        stream << std::showpos << CTEntry::val(edge.w.r) << CTEntry::val(edge.w.i) << std::noshowpos << "i\n";
+
+        std::unordered_set<decltype(edge.p)> nodes{};
+
+        std::priority_queue<const Edge*, std::vector<const Edge*>, priocmp> q;
+        q.push(&edge);
+
+        // bfs until finished
+        while (!q.empty()) {
+            auto edge_ptr = q.top();
```
Comment (burgholzer) [inline include/dd/Export.hpp:673]:
```suggestion
            auto edgePtr = q.top();
```

File: include/dd/Package.hpp [hunk 2]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+                p0 /= tmp;
+
+                const fp n = dist(mt);
+                if (n < p0) {
+                    cur = cur.p->e.at(0);
+                } else {
+                    result[cur.p->v] = '1';
+                    cur              = cur.p->e.at(1);
+                }
+            }
```
Comment (burgholzer) [inline include/dd/Package.hpp:719]: maybe use a more meaningful name for `n`, e.g., `threshold`?

File: include/dd/Package.hpp [hunk 3]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+
+            if (collapse) {
+                decRef(root_edge);
+
+                vEdge                e = vEdge::one;
+                std::array<vEdge, 2> edges{};
+
+                for (Qubit p = 0; p < n_qubits; p++) {
```
Comment (burgholzer) [inline include/dd/Package.hpp:727]:
```suggestion
                for (Qubit p = 0; p < nqubits; p++) {
```
`nqubits` or `nq`?

File: include/dd/Package.hpp [hunk 4]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+                incRef(e);
+                root_edge = e;
+                garbageCollect();
+            }
+
+            return std::string{result.rbegin(), result.rend()};
+        }
+
+    private:
+        double assignProbabilities(vEdge edge, std::unordered_map<vNode*, double>& probs) {
```
Comment (burgholzer) [inline include/dd/Package.hpp:746]:
```suggestion
        double assignProbabilities(const vEdge& edge, std::unordered_map<vNode*, double>& probs) {
```

File: include/dd/Package.hpp [hunk 5]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+                sum = assignProbabilities(edge.p->e.at(0), probs) + assignProbabilities(edge.p->e.at(1), probs);
+            }
+
+            probs.insert({edge.p, sum});
+
+            return ComplexNumbers::mag2(edge.w) * sum;
+        }
+
+    public:
+        char measureOneCollapsing(vEdge& root_edge, const Qubit index, const bool assume_probability_normalization, std::mt19937_64& mt, fp epsilon = 0.001) {
```
Comment (burgholzer) [inline include/dd/Package.hpp:764]:
```suggestion
        char measureOneCollapsing(vEdge& rootEdge, const Qubit index, const bool assumeProbabilityNormalization, std::mt19937_64& mt, fp epsilon = 0.001) {
```
Isn't probability normalisation always assumed?

File: include/dd/Package.hpp [hunk 6]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+            std::map<vNode*, fp> probsMone;
+            std::set<vNode*>     visited_nodes2;
```
Comment (burgholzer) [inline include/dd/Package.hpp:766]:
```suggestion
            std::set<vNode*>     visited;
```
?

File: include/dd/Package.hpp [hunk 7]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+            std::queue<vNode*>   q;
+
+            probsMone[root_edge.p] = ComplexNumbers::mag2(root_edge.w);
+            visited_nodes2.insert(root_edge.p);
+            q.push(root_edge.p);
+
+            while (q.front()->v != index) {
+                vNode* ptr = q.front();
+                q.pop();
+                double prob = probsMone[ptr];
```
Comment (burgholzer) [inline include/dd/Package.hpp:776]:
```suggestion
                fp prob = probsMone[ptr];
```

File: include/dd/Package.hpp [hunk 8]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+                    if (!ptr->e.at(0).w.approximatelyZero()) {
+                        pzero += probsMone[ptr] * ComplexNumbers::mag2(ptr->e.at(0).w);
+                    }
+
+                    if (!ptr->e.at(1).w.approximatelyZero()) {
+                        pone += probsMone[ptr] * ComplexNumbers::mag2(ptr->e.at(1).w);
+                    }
+                }
+            } else {
+                std::unordered_map<vNode*, double> probs;
```
Comment (burgholzer) [inline include/dd/Package.hpp:819]:
```suggestion
                std::unordered_map<vNode*, fp> probs;
```

File: include/dd/Package.hpp [hunk 9]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+                    }
+                }
+            }
+
+            if (std::abs(pzero + pone - 1) > epsilon) {
+                throw std::runtime_error("Numerical instability occurred during measurement: |alpha|^2 + |beta|^2 = " + std::to_string(pzero) + " + " + std::to_string(pone) + " = " +
+                                         std::to_string(pzero + pone) + ", but should be 1!");
+            }
+
+            const fp sum = pzero + pone;
```
Comment (burgholzer) [inline include/dd/Package.hpp:841]:
```suggestion
            const fp sum = pzero + pone;
            if (std::abs(sum - 1) > epsilon) {
                throw std::runtime_error("Numerical instability occurred during measurement: |alpha|^2 + |beta|^2 = " + std::to_string(pzero) + " + " + std::to_string(pone) + " = " +
                                         std::to_string(sum) + ", but should be 1!");
            }
```

File: include/dd/Package.hpp [hunk 10]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+
+            GateMatrix measure_m{
+                    complex_zero, complex_zero,
+                    complex_zero, complex_zero};
```
Comment (burgholzer) [inline include/dd/Package.hpp:845]:
```suggestion
            GateMatrix measurementMatrix{
                    complex_zero, complex_zero,
                    complex_zero, complex_zero};
```
Should `complex_zero` maybe also be called `complexZero`?

File: include/dd/Package.hpp (hunk 5 above)
Comment (burgholzer) [inline include/dd/Package.hpp:764]: Is `char` the right return type here? wouldn't `bool` be more suitable?

File: include/dd/Package.hpp [hunk 11]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+
+            std::uniform_real_distribution<fp> dist(0.0, 1.0L);
+
+            fp   n = dist(mt);
```
Comment (burgholzer) [inline include/dd/Package.hpp:849]: as above: more meaningful name?

File: include/dd/Package.hpp [hunk 12]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+            fp   norm_factor;
```
Comment (burgholzer) [inline include/dd/Package.hpp:850]:
```suggestion
            fp   normalizationFactor;
```

File: include/dd/Package.hpp [hunk 13]
```diff
@@ -676,6 +679,201 @@ namespace dd {
...
+                measure_m[0] = complex_one;
+                norm_factor  = pzero;
+                result       = '0';
+            } else {
+                measure_m[3] = complex_one;
+                norm_factor  = pone;
+                result       = '1';
+            }
+
+            mEdge m_gate = makeGateDD(measure_m, root_edge.p->v + 1, index);
```
Comment (burgholzer) [inline include/dd/Package.hpp:863]:
```suggestion
            auto measurementGate = makeGateDD(measure_m, root_edge.p->v + 1, index);
```

File: test/test_package.cpp [hunk 14]
```diff
@@ -732,3 +733,86 @@ TEST(DDPackageTest, Controls) {
...
     EXPECT_EQ(controls.count(0), 2);
 }
+
+TEST(DDPackageTest, DestructiveMeasurementAll) {
+    auto dd          = std::make_unique<dd::Package>(4);
+    auto h_gate0     = dd->makeGateDD(dd::Hmat, 2, 0);
+    auto h_gate1     = dd->makeGateDD(dd::Hmat, 2, 1);
+    auto plus_matrix = dd->multiply(h_gate0, h_gate1);
+    auto zero_state  = dd->makeZeroState(2);
+    auto plus_state  = dd->multiply(plus_matrix, zero_state);
```
Comment (burgholzer) [inline test/test_package.cpp:764]: Minor: camelCase?

File: test/test_package.cpp [hunk 15]
```diff
@@ -732,3 +733,86 @@ TEST(DDPackageTest, Controls) {
...
+    ASSERT_EQ(v_after[3], dd::complex_zero);
+
+    const auto v_after_compl = dd->getVectorStdComplex(plus_state);
+
+    assert(v_after.size() == v_after_compl.size());
+    for (std::size_t i = 0; i < v_after.size(); i++) {
+        ASSERT_DOUBLE_EQ(v_after.at(i).r, v_after_compl.at(i).real());
+        ASSERT_DOUBLE_EQ(v_after.at(i).i, v_after_compl.at(i).imag());
+    }
+}
```
Comment (burgholzer) [inline test/test_package.cpp:839]: Newline at end of file

**Review by burgholzer — CHANGES_REQUESTED** [review]
Overall this looks good. Mostly some naming convention stuff and a few minor questions.

File: include/dd/Package.hpp (hunk 5 above)
Comment (hillmich) [reply include/dd/Package.hpp:764]: Potentially, but `char` is a nice complement to the `std::string` returned by `measureAll`

**Review by burgholzer — APPROVED** [review]
LGTM 👍🏼

**Pull request merged**